### PR TITLE
PostgreSQL JSONB columns do not preserve the key order of hashes

### PR DESCRIPTION
**Pull request by bradleypriest**
### Summary

PostgreSQL JSONB columns do not preserve the key order of hashes

https://www.postgresql.org/docs/9.6/static/datatype-json.html

Due to this setting a JSONB column to the same hash twice can cause `model.changed?` to return false positives

### CODE REVIEW

**Comment by bradleypriest** [comment]
Sorry, just realized this breaks for arrays in JSONB columns, will ping once I have a more comprehensive fix

**Comment by bradleypriest** [comment]
Ok, it seems there's two possible solutions here:
a) Check for arrays and transform them separately - This updated PR (sorry for force-pushing, force of habit)
b) Compare the Ruby representations of the JSON objects instead of their serialized form, spike can be found here (https://github.com/rails/rails/compare/master...bradleypriest:jsonb-changed-ruby)

File: activerecord/CHANGELOG.md [hunk 1]
```diff
@@ -1,3 +1,6 @@
+*   Ignore key order when checking whether JSONB columns have changed in PostgreSQL
```
Comment (maclover7) [inline activerecord/CHANGELOG.md]: Can you add a blank line between the bullet point and your name?

**Review by maclover7 — CHANGES_REQUESTED** [review]
Hmm, would it make sense to use the same solution as #27517 and compare deserialized values (so instances of `Hash`) here? Equality through that route won't be affected by different key orders; this might be easier as opposed to sorting keys and doing the serialize/deserialize dance.

**Comment by bradleypriest** [comment]
@maclover7 thanks for pointing that out. As the logic was already repeated in the `hstore` and MySQL `json` types, I've just moved the logic up into the AbstractJSON type.

The only other type that inherits that is the Postgres JSON type, which shouldn't be detrimentally affected by this change in any way

File: activerecord/test/cases/adapters/postgresql/json_test.rb [hunk 2]
```diff
@@ -192,6 +192,20 @@ def test_changes_in_place
 
     assert_equal({ "one" => "two", "three" => "four" }, json.payload)
     assert_not json.changed?
+
+    json.payload = { "three" => "four", "one" => "two" }
```
Comment (maclover7) [inline activerecord/test/cases/adapters/postgresql/json_test.rb]: Can we move these tests to a more generic area / make sure the new behavior in `AbstractJson` is being tested properly?

Comment (bradleypriest) [inline activerecord/test/cases/adapters/postgresql/json_test.rb]: This file is already mixed into both JSON & JSONB test suites, it looks like the MySQL json_test file is mostly a copy-paste of this one.

I've added a couple more commits, the first adds the same test into the MySQL tests as well.
The second extracts out all the shared tests into a shared module for both MySQL and PostgreSQL

**Comment by bradleypriest** [comment]
My last comment got collapsed, but thanks to feedback from @maclover7 I've extracted out the shared JSON tests into a separate module to ensure both adapters using AbstractJSON are tested.

If that is too much for this PR, the extraction is isolated to just the last commit, so can be handled separately.

**Comment by ianks** [comment]
I'm confused @bradleypriest. The documentation states this:

> Because the json type stores an exact copy of the input text, it will preserve semantically-insignificant white space between tokens, as well as the order of keys within JSON objects.

**Comment by bradleypriest** [comment]
Hi @ianks, that's for the JSON type. For JSONB it continues 
> By contrast, jsonb does not preserve white space, does not preserve the order of object keys

**Comment by bradleypriest** [comment]
Rebased on the latest CHANGELOG conflicts, let me know if there's anything more I can do

**Comment by bradleypriest** [comment]
Rebased again, would love any feedback when someone has a chance.

**Comment by bradleypriest** [comment]
Rebased again.
@sgrif if there's anything I can do to help get some eyes on this, I'd love to help.

**Comment by jalberto** [comment]
Hi, In which release this patch has been included?

**Pull request closed without merging**